### PR TITLE
Add support for persistent backend

### PR DESCRIPTION
**Pull request by hobochili**
This is a rather large merge. Apologies that the commits are not more granular. I'm happy to expound on any changes and am open to critique or discussion about coding conventions.

Changes include:
- adds support for persistent game state via redis
- adds support for game state locking
- adds support for runtime locking (only impactful when using persistent state)
- adds graceful handling of SIGINT and SIGTERM
- adds requirements.txt for dependencies
- adds docker configs for app and redis
- introduces a new Config class and Dynaconf dependency for fluid config parsing from config.toml or .env
- adds basic logging
- moves Canvas reference outside of Game for ease of serialization
- adds `allow_channels` and `deny_channels` config options, useful if limiting testing of new changes to specific channels

### CODE REVIEW

**Comment by hobochili** [comment]
Just added a new ErrorHandler cog. At this point it would make sense to restructure the repo, I think, to something like:

image

One issue is that currently the ErrorHandler cog has to reach inside the Wordle cog to grab a reference for RedisConnectionError.

We could relocate some of the Bot setup from `Main.py` to `Bot/Bot.py`, too.

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+jedi==0.18.1
+jupyter-client==7.1.0
+jupyter-core==4.9.1
+matplotlib-inline==0.1.3
+multidict==5.2.0
+nest-asyncio==1.5.4
+parso==0.8.3
+pexpect==4.8.0
+pickleshare==0.7.5
+Pillow==8.4.0
```
Comment (RogueBurger) [inline requirements.txt:27]: I was having issue building the docker container due to Pillow being required in this file. I see earlier in the dockerfile that Pillow is installed along side some virtual dependencies. That installation worked, but then my guess, based on the errors I was seeing, is that the virtual dependencies went away by the time these requirements started getting installed, and so the Pillow install from this file failed for me. I went ahead and removed this Pillow requirement locally, and the build started working.

Comment (hobochili) [reply requirements.txt:27]: Hmm... I'll take a look. The Dockerfile was set up that way to limit the size of the resulting image. What system are you building on?

Comment (RogueBurger) [reply requirements.txt:27]: Windows

File: Wordle/GameManager.py [hunk 2]
```diff
@@ -1,25 +1,42 @@
...
+        self.store: Store = backend
+        self.canvas: Canvas = Canvas()
+
+    def create_game(self, word_length: int, mode: str) -> Game:
+        return Game(word_length=word_length, mode=mode, canvas=self.canvas)
+
+    @asynccontextmanager
+    async def lock(self, channel: TextChannel) -> Lock:
+        """ raises LockNotFoundError """
+        async with self.store.lock(channel.guild.id, channel.id):
```
Comment (RogueBurger) [inline Wordle/GameManager.py:22]: Referencing the `guild` throughout this file has broken the bot in DMs, as DMs are guildless channels. The bot working in DMs is a feature I'd like to keep if possible. Playing around with potential fixes, it looks like replacing each instance of `channel.guild.id` in the file with `channel.guild.id if not isinstance(channel, DMChannel) else 0` appears to work. That solution feels a wee bit hacky though, so gonna keep poking around for something a little cleaner.

File: requirements.txt (hunk 1 above)
Comment (hobochili) [reply requirements.txt:27]: I can't reproduce this on Linux (pop_OS). I just reset my venv, wiped my local docker images, and ran `docker-compose up -d --build` and it built fine. When exactly did it fail on your end? I wonder if in the hard-coded Pillow install it's pulling a different version than that declared in requirements.txt...

That said, it's not great how the Pillow install is hard-coded in Dockerfile. Can you try it again, restoring `requirements.txt` and swapping this in for your Dockerfile:

```
# syntax=docker/dockerfile:1

FROM python:3.9-alpine

WORKDIR /srv

COPY requirements.txt requirements.txt

RUN apk update \
    && apk add --virtual build-deps gcc python3-dev musl-dev \
    && apk add jpeg-dev zlib-dev libjpeg freetype-dev \
    && pip3 install -r requirements.txt \
    && apk del build-deps
```

Comment (RogueBurger) [reply requirements.txt:27]: New Dockerfile worked like a charm.

Old Dockerfile failed while building the wheel for Pillow.

Here's the original failure output:
```
...
#11 34.06   Building wheel for Pillow (setup.py): started
#11 34.41   Building wheel for Pillow (setup.py): finished with status 'error'
#11 34.41   ERROR: Command errored out with exit status 1:
#11 34.41    command: /usr/local/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-uznehizv/pillow_b1093849bbb14b3186f63f19448e2a4e/setup.py'"'"'; __file__='"'"'/tmp/pip-install-uznehizv/pi
llow_b1093849bbb14b3186f63f19448e2a4e/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"
'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-tql45hky
#11 34.41        cwd: /tmp/pip-install-uznehizv/pillow_b1093849bbb14b3186f63f19448e2a4e/
...
#11 34.41   error: command 'gcc' failed: No such file or directory
#11 34.41   ----------------------------------------
#11 34.41   ERROR: Failed building wheel for Pillow
...
```

File: Wordle/GameManager.py (hunk 2 above)
Comment (hobochili) [reply Wordle/GameManager.py]: Good catch.

File: requirements.txt (hunk 1 above)
Comment (hobochili) [reply requirements.txt:27]: Great, just pushed the new one.

**Pull request merged**